internal/configs/testdata/config-diagnostics/pass-inherited-provider/main.tf

```diff
@@ -5,3 +5,10 @@ provider "test" {
 module "mod" {
   source = "./mod"
 }
+
+# FIXME: This test case seems a little suspicious, because this module is
```

Comment:
This test is representative of a case that never worked correctly. Some older versions would use the wrong provider non-deterministically, and later they would return a "provider not found" error. As part of the overall provider work we decided that in order to reference a provider by name (rather than implicitly), that provider must be declared within the configuration.

Reply:
The trouble I was having with this particular requirement is that a seemingly-internal-only change to the child module can cause a warning to appear for an existing caller who was previously validly relying on the normal inheritance behavior, which makes this requirement hard to explain and justify in the warning message text: the audience of the warning is the author of the caller, but it was the author of the child module who caused the problem, probably unknowingly because this requirement is subtle and we don't have any way to proactively warn the child module author about it before publication.

I understand that it's a compatibility constraint and so I'm not proposing to make any changes to this behavior for the foreseeable future (perhaps in a future language edition we will require explicit provider requirements/ passing in all cases to avoid this) but I'd like to change this comment to describe the constraint you are describing for the benefit of future maintainers who might see a bug report for this inconsistency and want to determine why Terraform behaves this way.

I'm going to update this comment to capture something like what you have explained here, but otherwise leave this unchanged.